Standards/scs-0215-v1-robustness-features.md

```diff
@@ -0,0 +1,257 @@
+Possible flags, that can be set for this feature, are:
+
+* auto-compaction-mode
+* auto-compaction-retention
+
+More information about compaction can be found in the respective etcd documentation
+[etcd maintenance](https://etcd.io/docs/v3.3/op-guide/maintenance/)
+
+### etcd backup
+
+An etcd cluster should be regularly backed up in order to be able to restore the cluster to a known good state at an
+earlier space in time if a failure or incorrect state happens.
+The cluster should be backed up multiple times in order to have different possible states to go back to. This is especially
+useful, if data in the newer backups was already corrupted in some way or important data was deleted in them.
+For this reason, a backup strategy needs to be developed with a decreasing number of backups in an increasing period of time,
+meaning that the previous year should only have 1 backup, but the current week should have multiple.
+Information about the backup process can be found in the etcd documentation:
+[Upgrade etcd](https://kubernetes.io/docs/tasks/administer-cluster/configure-upgrade-etcd/)
+
+### Certificate rotation
+
+In order to secure the communication of a Kubernetes cluster, (TLS) certificates signed by a controlled
+Certificate Authority (CA) can be used.
+Normally, these certificates expire after a set period of time. In order to avoid expiration and failure of a cluster,
+these certificates need to be rotated regularly and at best automatically.
+It is important to either set `--rotate-server-certificates` as a command line parameter or set `rotateCertificates: true`
+in the kubelet config or the `kubeletExtraArgs` of the cluster-template.yaml file. This activates the rotation of the
+kubelet server certificate. Another recommendation is to set `serverTLSBootstrap: true`, which also enables the request
+and rotation of the certificate for the kubelet by requesting them from `certificates.k8s.io`
+according to the documentation under [Kubeadm certs].
+
+A clusters certificates can be rotated by updating the cluster, which according to the Kubernetes documentation
+automatically renews the certificates.
+Another method would be the manual renewal, which can be done through multiple methods, depending on the K8s cluster
+used. An example for a K8s cluster set up with `kubeadm` would be to execute the command
+
+```bash
+kubeadm certs renew all
+```
+
+Since clusters conformant with the SCS standards would probably be updated within the time period described in the
+standard [SCS-0210-v2](https://github.com/SovereignCloudStack/standards/tree/main/Standards/scs-0210-v2-k8s-version-policy.md),
+this rotation can probably be assumed to happen, because of the cluster update functionality.
+
+It is also important to mention, that due to security reasons, the CSR needs to be approved manually with the commands
+
+```bash
+kubectl get csr
+kubectl certificate approve <CSR>
+```
+
+in order to complete a certificate rotation.
```

Comment:
Related to https://github.com/SovereignCloudStack/standards/pull/361#discussion_r1448822622

> It is also important to mention, that due to security reasons, the CSR needs to be approved manually with the commands

Sure? https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-certs/#manual-certificate-renewal does not seem to mention it.

Reply:
Rewrote the section to hopefully fit better.

Reply:
Summoning @joshmue again to resolve (and maybe approve)

Reply:
The rewording from hard "needs to" etc. to "may need to" reduces the risk of strictly false statements, indeed.

My point (from https://github.com/SovereignCloudStack/standards/pull/361#discussion_r1447407033) still is that information should be verified with a high degree of certainty, which I do not feel there is:

> > About the CSR part, I'm not really sure. [...] But I'm really not an expert on this.
> >
> I did not dig into that to much right now, so I'm not totally sure whether this should be mandated here, either (I'd tend not to).
> If this is to be included and merged, someone should be sure about that, though.

Reply:
So I read up on most of the stuff again (https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-certs, https://kubernetes.io/docs/reference/access-authn-authz/kubelet-tls-bootstrapping, https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/) and came to the conclusion, to rewrite the `Certificate rotation` section, since there were quite a few errors and mixups in this part. IMO its now better structured and correct.

I also shortened the section related to this under `Decision` to the relevant stuff.

Reply:
Looks way better to me, as well! :+1: